fusedrug/utils/samplers/balanced_df_sampler.py

```diff
@@ -0,0 +1,110 @@
+from typing import List, Union, Any
+from torch.utils.data import Sampler
+import numpy as np
+import pandas as pd
+
+
+class BalancedClassDataFrameSampler(Sampler):
+    """
+    Balanced sampler for large dataframes.
+    (Native fuse BatchSamplerDefault caused memory issues when used with very large datasets).
+    :param df: the dataframe which stores the data
+    :param label_column_name: name of the label column in the dataframe
+    :param classes: list of class names to use
+    :param counts: list of number of times each class should be used in a minibatch. the minibatch size will be the sum of counts
+    :param shuffle: whether to shuffle  - both each minibatch, and once saw all samples from a class. TODO: consider separating into two params
+    :total_minibatches: total minibatches per epoch
+    """
+
+    def __init__(
+        self,
+        df: pd.DataFrame,
+        label_column_name: str,
+        classes: List[str],
+        counts: List[int],
+        shuffle: bool = True,
+        total_minibatches: Union[int, str] = "see_all_smallest_class",
+    ):
+        self.df = df
+        assert len(classes) == len(counts)
+        self.shuffle = shuffle
+        self.minibatch_size = sum(counts)
+        assert self.minibatch_size > 0
+
+        self.classes = classes
+        self.counts = counts
+
+        self.indices_per_class = []
+        self.pointers = []
+
+        minibatches_needed_to_see_all = []
+
+        for curr_class, curr_count in zip(self.classes, self.counts):
+            assert isinstance(curr_count, int)
+            indices = np.flatnonzero(df[label_column_name] == curr_class)
+            if len(indices) == 0:
+                raise Exception(f"found zero cases for class {curr_class} !")
+            else:
+                print(f"found {len(indices)} cases of class {curr_class}")
+            self.indices_per_class.append(indices)
+            self.pointers.append(indices.shape[0])
+
+            minibatches_needed_to_see_all.append(
+                int(np.ceil(indices.shape[0] / curr_count))
+            )
+
+        smallest_class_index = np.argmin([len(x) for x in self.indices_per_class])
```

Comment:
Shouldn't you look for the max in  minibatches_needed_to_see_all?

Reply:
every batch will contain elements from all classes, so the total number of minibatches is defined according to the rarest class. and then when there are no more examples in the rarest class, the "epoch" ends, and optionally reshuffles.
you could think of different logics, but I think this also makes sense. and there's a test for this sampler that verified the expected behavior